applications/tari_validator_node/src/p2p/services/mempool/service.rs

```diff
@@ -51,6 +54,39 @@ pub struct MempoolService {
     tx_valid_transactions: broadcast::Sender<(Transaction, ShardId)>,
     epoch_manager: EpochManagerHandle,
     node_identity: Arc<NodeIdentity>,
+    validator: MempoolTransactionValidator,
+}
+
+pub struct MempoolTransactionValidator {
+    template_manager: TemplateManager,
+}
+
+impl MempoolTransactionValidator {
+    pub(crate) fn new(template_manager: TemplateManager) -> Self {
+        Self { template_manager }
+    }
+}
+#[async_trait]
+impl Validator<Transaction> for MempoolTransactionValidator {
+    type Error = MempoolError;
+
+    async fn validate(&self, inner: &Transaction) -> Result<(), MempoolError> {
+        let instructions = inner.instructions();
+
+        for instruction in instructions {
+            match instruction {
+                Instruction::CallFunction { template_address, .. } => {
+                    match self.template_manager.fetch_template(template_address) {
```

Comment:
nit(perf): Should probably add a `template_manager.exists(template_address)` call so that it doesn't have to fetch the template